src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1109,17 +1114,22 @@ void HNSWIndex<DataType, DistType>::replaceEntryPoint() {
 }
 
 template <typename DataType, typename DistType>
+template <bool has_marked_deleted>
 void HNSWIndex<DataType, DistType>::SwapLastIdWithDeletedId(idType element_internal_id) {
-    // swap label
-    replaceIdOfLabel(getExternalLabel(cur_element_count), element_internal_id, cur_element_count);
+    // Swap label - this is relevant when the last element's label exists (it is not marked as
+    // deleted). For inplace delete, this is always the case.
+    if (!has_marked_deleted || !isMarkedDeleted(cur_element_count)) {
```

Comment:
no... but if `has_marked_deleted` we need to check other condition

Reply:
so !isMarkedDeleted(cur_element_count) is enoguh for both cases

Reply:
Oh right, but we do it like this because the compiler wouldn't even generate the code if we are running in non-tiered mode (in which case `has_marked_deleted` is evaluated as false in compile time)